functions/index.js

```diff
@@ -0,0 +1,277 @@
+admin.initializeApp(functions.config().firebase);
+
+// Create and Deploy Your First Cloud Functions
+// OFFICIAL DOCUMENTATION
+// https://firebase.google.com/docs/functions/write-firebase-functions
+
+
+// THIS FUNCTION ADDS PUSHTOKEN FOR USERID
+exports.addPushTokenForUserId = functions.https.onRequest((req, res) => {
+
+  if (req.get('FUNCTION_SECRET_KEY') !== functions.config().functions.secret) {
+    console.log('not authenticated, FUNCTION_SECRET_KEY header is barps');
+    res.sendStatus(403);
+    return;
+  }
+
+  const userId = req.query.userId;
+  const pushToken = req.query.pushToken;
+
+  admin.database().ref('/pushTokens/' + userId).set({token: pushToken}).then(snapshot => {
+    console.log('pushToken added for userId ' + userId);
+    res.sendStatus(200);
+  });
+});
+
+// this function takes an userId as an argument and removes that user's
+// pushToken from the database. after this the user will not receive any
+// push notifications from the service anymore.
+exports.removeUserId = functions.https.onRequest((req, res) => {
+
+  if (req.get('FUNCTION_SECRET_KEY') !== functions.config().functions.secret) {
+    console.log('not authenticated, FUNCTION_SECRET_KEY header is barps');
+    res.sendStatus(403);
+    return;
+  }
+
+  const userId = req.query.userId;
+
+  return admin.database().ref('/pushTokens/' + userId).set({ token: null }).then(snapshot => {
+    console.log('pushToken removed from userId ' + userId);
+    // return admin.database().ref('/chats/').once('value').then(snapshot => {
+    //   const allChats = snapshot.val();
+    //   const chatsToClose = [];
+    //   Object.keys(allChats).forEach(key => {
+    //     if (allChats[key].users[0] == userId || allChats[key].users[1] == userId) {
+    //       chatsToClose.push(key);
+    //     }
+    //   });
+    //   // these are the keys for chats of this userId
+    //   // TODO: it should be decided what to do with them
+    //   // leave open, close them?
+    //   // should the other user be notified somehow?
+    //   console.log('chatsToClose:');
+    //   console.log(chatsToClose);
+    // });
+    res.sendStatus(200);
+  });
+
+});
+
+// this function adds a new chat between two users
+// the chat will be in /messages and it will have a random key
+// there will be an initial message in the chat which can be tuned below
+exports.addNewChatBetweenUsers = functions.https.onRequest((req, res) => {
+
+  if (req.get('FUNCTION_SECRET_KEY') !== functions.config().functions.secret) {
+    console.log('not authenticated, FUNCTION_SECRET_KEY header is barps');
+    res.sendStatus(403);
+    return;
+  }
+
+  console.log(req.query)
+  const userId1 = req.query.userId1;
+  const userId2 = req.query.userId2;
+
+  return admin.database().ref('/chats/')
+         .push({
+           messages: [],
+           users: [userId1, userId2]
+         })
+    .then(snapshot => {
+      console.log('chat record was added for users ' + userId1 + ' and ' + userId2);
```

Comment:
This could be replaced with a template string like so:

```javascript
console.log(`chat record was added for users ${userId1} and ${userId2}`);
```